In [1]:
#Imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import duckdb
import sqlalchemy
#%load_ext sql

In [2]:
%config SqlMagic.autopandas = True
%config SqlMagic.feedback = False
%config SqlMagic.displaycon = False

**Task: Automated Ontology Refinement with Sentence Similarity**
<br>
- Task
  - Identification of new entity candidates
  - Semi-automatic procedure for enhancing the ontology
- Learning
  - SentenceBERT (pretrained)
    - Fine-tuning with ontology
    - Pairs of phrases, e.g. entities and synonyms
    - Identification of candidate entities (noun phrases?)
- Input
  - entities.csv
  - sentences.csv


# Test iteration over sample article and find similarities between ontology and words in article 

## Load data

### `sentences.csv`

In [3]:
# Load sentences
sentences = pd.read_csv("../Data/FulltextofCorpus/sentences.csv")

In [4]:
# Assess shape of sentences.csv
sentences.shape

(6563319, 10)

In [5]:
# Subset only first article
sample_article_df = sentences.loc[sentences.article_id==1]
sample_article_df.shape

(1013, 10)

In [6]:
sample_article_df.head()

article_id sentence_id para_id  \
0           1       1_0_1     1_0   
1           1       1_1_2     1_0   
2           1      1_2_22     1_0   
3           1     1_22_23     1_1   
4           1     1_23_24     1_1   

                                            sentence  section_nr  \
0                                                              0   
1                                    START_TITLE_TAG           0   
2  Examining interdependence between product user...           0   
3                                 START_ABSTRACT_TAG           0   
4                                                              0   

  last_section_title last_subsection_title sentence_type references  \
0                NaN                   NaN         EMPTY         []   
1                NaN                   NaN           TAG         []   
2                NaN                   NaN     PARAGRAPH         []   
3                NaN                   NaN           TAG         []   
4           Abstract                   NaN         EMPTY         []   

   number_of_citations  
0                    0  
1                    0  
2                    0  
3                    0  
4                    0

In [7]:
# Check values in colums of interest
sample_article_df.sentence_type.value_counts()

PARAGRAPH        357
TAG              279
EMPTY            146
TABLE            135
HEADER            24
CAPTION           22
FORMULA           17
ABSTRACT          11
ANNEX             11
TABLE_HEADER       6
FIGURE_HEADER      2
HYP_NUMBER         1
HYPOTHESIS         1
FIGURE             1
Name: sentence_type, dtype: int64

### `isrecon.duckdb`

In [13]:
# Note that DBeaver must be closed, before you can run this command!
con = duckdb.connect(database='../Data/ExtractedInformationofCorpus/isrecon.duckdb', read_only=False)

In [57]:
# Load first article as df
entities_sample_article = entities = con.execute("""
                       SELECT *
                       FROM entities e
                       WHERE e.article_id = 1""").fetchdf()

entities_sample_article.shape

(132, 46)

In [58]:
entities_sample_article.ent_id.nunique()

46

In the first article 
- Total of 132 entities have been found
- 46 unique entities have been found

In [19]:
entities_sample_article.head()

article_id para_id  sentence_start sentence_id section_title  \
0           1     1_3             322   1_322_354  Introduction   
1           1     1_3             467   1_467_502  Introduction   
2           1     1_3             502   1_502_527  Introduction   
3           1     1_4             530   1_530_551  Introduction   
4           1     1_5             861   1_861_886  Introduction   

  subsection_title         label                ent_id  \
0              NaN         TOPIC  knowledge repository   
1              NaN         TOPIC      digital platform   
2              NaN    TECHNOLOGY        Salesforce.com   
3              NaN         TOPIC  participatory design   
4              NaN  PARTICIPANTS     group participant   

                  level_1        level_2  ... attr_double attr_stage  \
0  domain specific entity       IS topic  ...           0          0   
1  domain specific entity       IS topic  ...           0          0   
2  domain specific entity  IS technology  ...           0          0   
3  domain specific entity       IS topic  ...           0          0   
4  domain specific entity   study object  ...           0          0   

  attr_type attr_temporal attr_assumption attr_dsr attr_paired attr_setting  \
0         0             0               0        0           0            0   
1         0             0               0        0           0            0   
2         0             0               0        0           0            0   
3         0             0               0        0           0            0   
4         0             0               0        0           0            0   

  attr_level attr_tool  
0          0         0  
1          0         0  
2          0         0  
3          0         0  
4          1         0  

[5 rows x 46 columns]

In [55]:
#Check columns and datatypes
entities_sample_article.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Data columns (total 46 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   article_id        132 non-null    int64  
 1   para_id           132 non-null    object 
 2   sentence_start    132 non-null    int64  
 3   sentence_id       132 non-null    object 
 4   section_title     120 non-null    object 
 5   subsection_title  0 non-null      object 
 6   label             132 non-null    object 
 7   ent_id            132 non-null    object 
 8   level_1           132 non-null    object 
 9   level_2           132 non-null    object 
 10  level_3           130 non-null    object 
 11  level_4           118 non-null    object 
 12  level_5           66 non-null     object 
 13  level_6           19 non-null     object 
 14  level_7           6 non-null      object 
 15  level_8           3 non-null      object 
 16  level_9           2 non-null      object 
 1

- It needs to be clarified, which of these columns are relevant for the task (assumption only first ~17 are somehow relevant for us)
- Which columns do we need for the training (probably only sentence_id, entity_id, synoyms)? 

In [79]:
# Check which sentences have more than one label 
entities_sample_article[entities_sample_article['sentence_id']\
                        .map(entities_sample_article['sentence_id'].value_counts()) > 1]

article_id para_id  sentence_start    sentence_id  \
11            1    1_12            1410    1_1410_1433   
12            1    1_12            1410    1_1410_1433   
14            1    1_12            1455    1_1455_1476   
15            1    1_12            1455    1_1455_1476   
16            1    1_12            1476    1_1476_1498   
17            1    1_12            1476    1_1476_1498   
19            1    1_19            2348    1_2348_2387   
20            1    1_19            2348    1_2348_2387   
23            1    1_19            2502    1_2502_2568   
24            1    1_19            2502    1_2502_2568   
35            1    1_28            3572    1_3572_3609   
36            1    1_28            3572    1_3572_3609   
38            1    1_28            3713    1_3713_3755   
39            1    1_28            3713    1_3713_3755   
40            1    1_28            3713    1_3713_3755   
41            1    1_28            3713    1_3713_3755   
46            1    1_32            4347    1_4347_4361   
47            1    1_32            4347    1_4347_4361   
72            1    1_66            6816    1_6816_6843   
73            1    1_66            6816    1_6816_6843   
92            1    1_80            7989    1_7989_8022   
93            1    1_80            7989    1_7989_8022   
94            1    1_80            7989    1_7989_8022   
95            1    1_80            8051    1_8051_8110   
96            1    1_80            8051    1_8051_8110   
99            1    1_80            8187    1_8187_8224   
100           1    1_80            8187    1_8187_8224   
107           1    1_93            9812    1_9812_9873   
108           1    1_93            9812    1_9812_9873   
109           1    1_93            9873    1_9873_9914   
110           1    1_93            9873    1_9873_9914   
117           1   1_111           10835  1_10835_10877   
118           1   1_111           10835  1_10835_10877   
126           1   1_131           11845  1_11845_11883   
127           1   1_131           11845  1_11845_11883   
128           1   1_131           11845  1_11845_11883   
129           1   1_131           11845  1_11845_11883   

                                         section_title subsection_title  \
11             Firm-hosted online branding communities              NaN   
12             Firm-hosted online branding communities              NaN   
14             Firm-hosted online branding communities              NaN   
15             Firm-hosted online branding communities              NaN   
16             Firm-hosted online branding communities              NaN   
17             Firm-hosted online branding communities              NaN   
19   Social exchange theory and knowledge contribut...              NaN   
20   Social exchange theory and knowledge contribut...              NaN   
23   Social exchange theory and knowledge contribut...              NaN   
24   Social exchange theory and knowledge contribut...              NaN   
35   Reading Employee-Generated content and knowled...              NaN   
36   Reading Employee-Generated content and knowled...              NaN   
38   Reading Employee-Generated content and knowled...              NaN   
39   Reading Employee-Generated content and knowled...              NaN   
40   Reading Employee-Generated content and knowled...              NaN   
41   Reading Employee-Generated content and knowled...              NaN   
46   Readership of Employee-Generated content and k...              NaN   
47   Readership of Employee-Generated content and k...              NaN   
72                      Discussion quantity (Quantity)              NaN   
73                      Discussion quantity (Quantity)              NaN   
92                                             Table 5              NaN   
93                                             Table 5              NaN   
94                                             Table 5              NaN   


Each sentence in entities can have more than one entities -> this will lead to an increase in rows when merged to sentences.csv

### `isntology.csv`

In [24]:
# Load ontology
ontology_syn = pd.read_csv("../Data/SynonymsinOntology/isontology.csv", index_col=0)
ontology_syn.shape

(390838, 4)

In [25]:
#Display head
ontology_syn.head()

entity_id                category  label  \
0     IS topic  domain specific entity  TOPIC   
1     IS topic  domain specific entity  TOPIC   
2  IS sourcing  domain specific entity  TOPIC   
3  IS sourcing  domain specific entity  TOPIC   
4  IS sourcing  domain specific entity  TOPIC   

                                             synonym  
0                                          is topics  
1                                           is topic  
2  information and communications technologies so...  
3                                        it sourcing  
4     information communications technology sourcing

In [34]:
# Check different values for "label" column
ontology_syn.label.value_counts()

PARTICIPANTS         235200
COLLECTION_METHOD     60176
TECHNOLOGY            39336
TOPIC                 22599
VALIDITY              16874
ANALYSIS_METHOD        4051
CONCEPTUAL_METHOD      3219
SECTOR                 2869
METRIC                 2728
MODEL_ELEMENT          1302
THEORY                 1188
COMPANY_TYPE            550
SAMPLING                372
PARADIGM                330
LEVEL                    44
Name: label, dtype: int64

- Looking at the ontology in protege, it is not clear for me what exactly the label column contains and how it is related to the hierarchy of the ontology. 
- My guess is that it is the first level of the ontology, but it seems that the words in the .csv slightly differ from the ones in protege or .json (e.g. json: "research paradigm" but in isontology.csv: "PARADIGM"

In [41]:
# Check different values for "category" column
ontology_syn.category.value_counts()

study object              235750
domain specific entity     64804
research method            63395
methodological entity      24025
theoretical entity          2864
Name: category, dtype: int64

- The column `category` seem to contain the "level 0" of the ontology 
- However this level 0 does no texist in the json file and in protege the five different categories are distributed over the 1st & 2nd level ? (Only domain specific entity, methodological entity, theoretical entity are in 1st level of protege file)

In [51]:
# Check number of synonyms per entity_id
ontology_syn.groupby("entity_id")[["synonym"]].nunique().sort_values(by="synonym", ascending=False)

synonym
entity_id                                    
individual participant                 220966
survey                                  42098
three validities                        16002
online learning system                  10075
communication service infrastructure     8932
...                                       ...
generative adversarial network              2
mutual information                          1
World of Warcraft                           1
conditional mutual information              1
Google news                                 1

[2689 rows x 1 columns]

For each of the 2689 unique entities there is one/many synonym(s)

In [59]:
# Apparently there are 220_966 synoyms for individual participant
ontology_syn[ontology_syn.entity_id=="individual participant"]

entity_id      category         label  \
26      individual participant  study object  PARTICIPANTS   
27      individual participant  study object  PARTICIPANTS   
28      individual participant  study object  PARTICIPANTS   
29      individual participant  study object  PARTICIPANTS   
30      individual participant  study object  PARTICIPANTS   
...                        ...           ...           ...   
220987  individual participant  study object  PARTICIPANTS   
220988  individual participant  study object  PARTICIPANTS   
220989  individual participant  study object  PARTICIPANTS   
220990  individual participant  study object  PARTICIPANTS   
220991  individual participant  study object  PARTICIPANTS   

                                                  synonym  
26      $num? $num multicultural and professional cust...  
27          $num? $num is and multicultural senior client  
28      $num? $num senior and business professional pr...  
29      $num? $num information and multicultural senio...  
30      $num? $num professional and executive knowledg...  
...                                                   ...  
220987  $num? $num business and senior financial profe...  
220988  $num? $num financial and professional business...  
220989  $num? $num business and it multicultural emplo...  
220990                security business citizens ( $num )  
220991     $num? $num chief and usable knowledge officers  

[220966 rows x 4 columns]

## Merge tables

In [82]:
# Merge sentences from first article with entities
sample_art_ent = sample_article_df.\
    merge(entities_sample_article[["sentence_id", "ent_id"]], how="left", on="sentence_id")

In [83]:
# Compare snumber of rows from sentences with merged df
sample_art_ent.shape[0]-sample_article_df.shape[0]

21

There are 21 "new" rows, since one sentence can have multiple entities

In [84]:
# Show some sentences with multiple entities
sample_art_ent[sample_art_ent['sentence_id'].map(sample_art_ent['sentence_id'].value_counts()) > 1].head()

article_id  sentence_id para_id  \
89           1  1_1410_1433    1_12   
90           1  1_1410_1433    1_12   
92           1  1_1455_1476    1_12   
93           1  1_1455_1476    1_12   
94           1  1_1476_1498    1_12   

                                             sentence  section_nr  \
89  We next review extant studies on OUC and devel...           2   
90  We next review extant studies on OUC and devel...           2   
92  We collected a total of 815 community document...           2   
93  We collected a total of 815 community document...           2   
94  In addition, a total of 12,315 product users w...           2   

                         last_section_title last_subsection_title  \
89  Firm-hosted online branding communities                   NaN   
90  Firm-hosted online branding communities                   NaN   
92  Firm-hosted online branding communities                   NaN   
93  Firm-hosted online branding communities                   NaN   
94  Firm-hosted online branding communities                   NaN   

   sentence_type references  number_of_citations                  ent_id  
89     PARAGRAPH         []                    0     research hypothesis  
90     PARAGRAPH         []                    0  social exchange theory  
92     PARAGRAPH         []                    0       group participant  
93     PARAGRAPH         []                    0  individual participant  
94     PARAGRAPH         []                    0  individual participant

In [95]:
# Left Join synonyms on merged_df
sample_art_ent_syn = sample_art_ent.merge(ontology_syn, how="left", left_on="ent_id", right_on="entity_id")

In [96]:
# Check shape 
sample_art_ent_syn.shape

(1581937, 15)

In [100]:
# Show number of synonyms/ rows per sentence & entity
sample_art_ent_syn.groupby(["sentence_id", "entity_id"])[["synonym"]].nunique()\
    .sort_values(by="synonym", ascending=False)

synonym
sentence_id   entity_id                      
1_1455_1476   individual participant   220966
1_4457_4496   individual participant   220966
1_5045_5074   individual participant   220966
1_5074_5085   individual participant   220966
1_2641_2667   individual participant   220966
...                                       ...
1_3835_3866   blog                          2
1_4425_4457   competitive advantage         2
1_10695_10701 Salesforce.com                2
1_4701_4727   productivity                  2
1_9873_9914   knowledge creation            2

[129 rows x 1 columns]

The shape of the dataframe increased dramatically, because some entities have a high number of synonyms and thus "blow" up the dataframe

In [ ]:
# Remove empty and tag sentences
sample_art_ent_syn_clean = sample_art_ent_syn.loc[~sample_art_ent_syn.sentence_type.isin(["EMPTY", "TAG"])]

## Load pretrained Sentence-BERT

In [101]:
#!pip install -U sentence-transformers

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.7/79.7 kB 1.2 MB/s eta 0:00:00a 0:00:01
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.2/4.2 MB 12.3 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.9/129.9 MB 6.2 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 18.3 MB/s eta 0:00:0000:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 20.2 MB/s eta 0:00:0000:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.2/86.2 kB 3.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.6/3.6 MB 22.9 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 21.3 MB/s eta 0:00:0000:0100:01
  Created wheel for sentence-transformers: filename=sentence_transformers-2.2.0-py3-none-any.whl size=120747 sha256=a88be8c6d679d2717b3d5899ef8bb4eafbb1a610797b488f97dbf1a20cceaf49
  Stored in directory: /Users/philippheitmann/Library/Caches/pip/wheels/0c/b6/fb/22

In [102]:
# Import sentence_transformers 
from sentence_transformers import SentenceTransformer, util

In [103]:
# Load a pre-trained model
model = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')

Downloading:   0%|          | 0.00/1.18k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/190 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/10.2k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/612 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/116 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/39.3k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/349 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/112 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/466k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/350 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/13.2k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/232k [00:00<?, ?B/s]

### Test SBERT & Create functions for similarity comparison (taken from `MiningRelationships.ipynb`)

In [105]:
# #Test embeddings
# sentences = ["This is an example sentence", "Each sentence is converted"]
# embeddings = model.encode(sentences)
# print(embeddings)

In [106]:
# Use the model to find pair similarities
def find_similarity(model, term1, term2):

    # Compute embedding for both lists
    embeddings1 = model.encode([term1], convert_to_tensor=True)
    embeddings2 = model.encode([term2], convert_to_tensor=True)

    # Compute cosine-similarits
    cosine_score = util.pytorch_cos_sim(embeddings1, embeddings2).item()

    print(cosine_score)

In [107]:
find_similarity(model,"software developer", "software engineer")
find_similarity(model,"database designer", "data architect")
find_similarity(model,"welder", "solderer")

0.8729231953620911
0.6365391612052917
0.32684826850891113


In [155]:
def find_top_similar(model, terms, k=10):
    '''
    Use the model to find similar pairs (compare all sentences against all other sentences)
    Returns a list with the pairs that have the highest cosine similarity score -> [score, idx_sentence1, idx_sentence2]     
    '''
    # Compute highest cosine similarity score
    paraphrases = util.paraphrase_mining(model, terms, show_progress_bar = True, top_k=k)
    
    # Store result in df & add sentences as text
    tmp_df = pd.DataFrame(paraphrases, columns=["cos_sim_score", "idx_sent1", "idx_sent2"])
    tmp_df[["txt_sent1", "txt_sent2"]] = tmp_df[["idx_sent1", "idx_sent2"]].applymap(lambda x:  terms[x])
    
    return tmp_df

In [156]:
find_top_similar(model, ["data science", "data engineer", "welding", "soldering"], 1)

Batches:   0%|          | 0/1 [00:00<?, ?it/s]

cos_sim_score  idx_sent1  idx_sent2     txt_sent1      txt_sent2
0       0.684671          0          1  data science  data engineer
1       0.552656          2          3       welding      soldering

### Find similarities between different sentences in sample article 

In [126]:
# Remove empty and tag sentences from article
sample_article_df_clean = sample_article_df.loc[~sample_article_df.sentence_type.isin(["EMPTY", "TAG"])]

In [137]:
all_sentences = sample_article_df_clean.sentence.tolist()

In [157]:
# Find similar sentences within smaple article
sim_sentences = find_top_similar(model, all_sentences, 2)

Batches:   0%|          | 0/19 [00:00<?, ?it/s]

In [162]:
sim_sentences

cos_sim_score  idx_sent1  idx_sent2  \
0         1.000001        305        418   
1         1.000000        152        449   
2         1.000000        291        410   
3         1.000000        291        420   
4         1.000000        303        416   
..             ...        ...        ...   
889       0.297382        587        558   
890       0.290877        573        151   
891       0.263151        395        288   
892       0.260414        389        385   
893       0.255322        573        587   

                                        txt_sent1  \
0    Values less than 0.001 are rounded to 0.001.   
1                                               4   
2                                               1   
3                                               1   
4                             at p < 0.01; * sig.   
..                                            ...   
889                                        Weifei   
890        Finding \nDholakia et al. \n \n \n \n    
891                                 Wiertz and De   
892                                  Jeppesen and   
893        Finding \nDholakia et al. \n \n \n \n    

                                             txt_sent2  
0         Values less than 0.001 are rounded to 0.001.  
1                                                    4  
2                                                    1  
3                                                    1  
4                                  at p < 0.01; * sig.  
..                                                 ...  
889                                     Hausman \n8117  
890  (e.g.,  START_CITE Singh et al., 2014; END_CIT...  
891                exp(γ'Zi (t-1)) 1 + exp(γ'Zi (t-1))  
892                                 Baron (2007, 2009)  
893                                             Weifei  

[894 rows x 5 columns]

- There are a lot of identical sentences in the article. 
- However, since our task is not to compare similarity between sentences we ignore this for now
- It might be useful to drop duplicate sentences and maybe preprocess sentences before training/ similarity comparison?

### Find similarities between ontology (without synonyms) and sentences in sample article 

In [165]:
#List of unique entities from ontology (first without synonyms)
entities_unique = ontology_syn.entity_id.unique().tolist()

In [167]:
#Record execution time
%%time

#Compute embeddings
embeddings_sentences = model.encode(all_sentences, convert_to_tensor=True)
embeddings_entities = model.encode(entities_unique, convert_to_tensor=True)

#Compute cosine-similarities for each sentence with each other sentence
cosine_scores = util.cos_sim(embeddings_sentences, embeddings_entities)

#Find the pairs with the highest cosine similarity scores
pairs = []
for i in range(len(cosine_scores)-1):
    for j in range(i+1, len(cosine_scores)):
        pairs.append({'index': [i, j], 'score': cosine_scores[i][j]})

#Sort scores in decreasing order
pairs = sorted(pairs, key=lambda x: x['score'], reverse=True)

for pair in pairs[0:10]:
    i, j = pair['index']
    print("{} \t\t {} \t\t Score: {:.4f}".format(all_sentences[i], entities_unique[j], pair['score']))

This is an example sentence 		 Each sentence is converted 		 Score: 0.1306
